helm/cluster-resources/templates/prometheus-get-metrics-rbac.yaml

```diff
@@ -0,0 +1,29 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: {{ include "resource.prometheus-get-metrics.name" . }}
+  labels:
+    {{- include "labels.common" . | nindent 4 }}
+rules:
+  - nonResourceURLs:
+      - "/metrics"
+    verbs:
+      - get
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: {{ include "resource.prometheus-get-metrics.name" . }}
+  labels:
+    {{- include "labels.common" . | nindent 4 }}
+subjects:
+  # prometheus is accessing kube-controller-manager and kube-scheduler /metrics
+  # endpoint using apiserver proxy which in turn makes anonymous requests to
+  # pods.
+  - kind: User
+    name: system:anonymous
```

Comment:
Our prometheus is using apiserver proxy to scrape metrics because it (usually, apart from the MC) runs outside of the cluster. Apparently when apiserver forwards the request it does it as `system:anonymous`. This is basically the same as running `curl` from a pod.

But there is no upstream requirement for that.

Here's a cluster API with those manifests deployed: https://216.119.153.248:6443. And you're right `curl -k https://216.119.153.248:6443/metrics` prints metrics :(

Reply:
But it doesn't escalate it further:

```
$ curl -k https://216.119.153.248:6443/api/v1/namespaces/kube-system/pods/https:kube-controller-manager-pawe2-control-plane-7kb7n-cvt7n:10257/proxy/metrics
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {

  },
  "status": "Failure",
  "message": "pods \"https:kube-controller-manager-pawe2-control-plane-7kb7n-cvt7n:10257\" is forbidden: User \"system:anonymous\" cannot get resource \"pods/proxy\" in API group \"\" in the namespace \"kube-system\"",
  "reason": "Forbidden",
  "details": {
    "name": "https:kube-controller-manager-pawe2-control-plane-7kb7n-cvt7n:10257",
    "kind": "pods"
  },
  "code": 403
}%                                                                                                                                                                                                                               
```